abdiff/core/run_ab_transforms.py

```diff
@@ -78,72 +75,101 @@ def run_ab_transforms(
         f"{[transformed_directory_a, transformed_directory_b]}"
     )
-
     run_configs = [
         (image_tag_a, transformed_directory_a),
         (image_tag_b, transformed_directory_b),
     ]
 
-    containers = []
-    for input_file in input_files:
-        source, output_file = parse_transform_details_from_extract_filename(input_file)
-        for docker_image, transformed_directory in run_configs:
-            container = run_docker_container(
-                docker_image,
-                transformed_directory,
-                source,
-                input_file,
-                output_file,
-                docker_client,
-            )
-            containers.append(container)
-
-    failed_containers = wait_for_containers(containers, timeout)
-    logger.info(f"All {len(containers)} containers have exited.")
+    # run containers and collect results
+    futures = run_all_docker_containers(docker_client, input_files, run_configs)
+    containers, exceptions = collect_container_results(futures)
+    logger.info(
+        f"Successful containers: {len(containers)}, failed containers: {len(exceptions)}"
+    )
 
+    # process results
     log_file = aggregate_logs(run_directory, containers)
     logger.info(f"Log file created: {log_file}")
-
-    # errors from failed containers are accessible via aggregated logs
-    if failed_containers:
-        raise DockerContainerRunFailedError(failed_containers)
-
+    if exceptions:
+        raise RuntimeError(  # noqa: TRY003
+            f"{len(exceptions)} / {len(containers) + len(exceptions)} containers failed "
+            "to complete successfully."
+        )
     ab_transformed_file_lists = get_transformed_files(run_directory)
-
     validate_output(ab_transformed_file_lists, len(input_files))
 
+    # write and return results
     run_data = {
         "input_files": input_files,
         "transformed_files": ab_transformed_file_lists,
     }
     update_or_create_run_json(run_directory, run_data)
-
     elapsed_time = perf_counter() - start_time
     logger.info(
         "Total time to complete process: %s", str(timedelta(seconds=elapsed_time))
     )
     return ab_transformed_file_lists
 
 
+def run_all_docker_containers(
+    docker_client: docker.client.DockerClient,
+    input_files: list[str],
+    run_configs: list[tuple],
+) -> list[Future]:
+    """Invoke Docker containers to run in parallel via threads.
+
+    By default, when ThreadPoolExecutor is invoked via a context manager it will wait for
+    all tasks to complete before exiting the context manager.  While each container is run
+    in a detached mode, the function run_docker_container() waits for the container to
+    exit making it effectively blocking.  The net result: this function will run all
+    containers to completion, returning the completed tasks (Future objects) as a list.
+    """
+    tasks = []
+
+    with ThreadPoolExecutor(max_workers=CONFIG.transmogrifier_max_workers) as executor:
```

Comment:
Why not utilize the `timeout` option for the threads to establish how long a Container can run?  

Because we run them in a detached state, even if the thread is terminated via a timeout, the container would continue running.  This is why `run_docker_container()` contains the logic to see how long a container has been running, and if exceeding a timeout, explicitly calls `container.stop()` to terminate it.